modules/core/04-channel/keeper/upgrade_test.go

```diff
@@ -126,6 +131,185 @@ func (suite *KeeperTestSuite) TestChanUpgradeInit() {
+			func() {
+				conn := path.EndpointB.GetConnection()
+				conn.State = connectiontypes.INIT
+				path.EndpointB.SetConnection(conn)
+			},
+			connectiontypes.ErrInvalidConnectionState,
+		},
+		{
+			"failed verification for counterparty channel state due to incorrectly constructed counterparty channel",
+			func() {
+				counterpartyChannel.State = types.CLOSED
+			},
+			commitmenttypes.ErrInvalidProof,
+		},
+		{
+			"failed verification for counterparty upgrade due to incorrectly constructed counterparty upgrade",
+			func() {
+				counterpartyUpgrade.LatestSequenceSend = 100
+			},
+			commitmenttypes.ErrInvalidProof,
+		},
+		{
+			"upgrade sequence mismatch, endpointB channel upgrade sequence is ahead",
+			func() {
+				channel := path.EndpointB.GetChannel()
+				channel.UpgradeSequence++
+				path.EndpointB.SetChannel(channel)
+			},
+			types.ErrIncompatibleCounterpartyUpgrade,
+		},
+		{
+			"upgrade ordering is not the same on both sides",
+			func() {
+				upgrade.Fields.Ordering = types.ORDERED
+			},
+			types.ErrIncompatibleCounterpartyUpgrade,
+		},
+		{
+			"proposed connection is not found",
+			func() {
+				upgrade.Fields.ConnectionHops[0] = invalidConnection
+			},
+			connectiontypes.ErrConnectionNotFound,
+		},
+		{
+			"proposed connection is not in OPEN state",
+			func() {
+				// reuse existing connection to create a new connection in a non OPEN state
+				connectionEnd := path.EndpointB.GetConnection()
+				connectionEnd.State = connectiontypes.UNINITIALIZED
+				connectionEnd.Counterparty.ConnectionId = counterpartyUpgrade.Fields.ConnectionHops[0] // both sides must be each other's counterparty
+
+				// set proposed connection in state
+				proposedConnectionID := "connection-100"
+				suite.chainB.GetSimApp().GetIBCKeeper().ConnectionKeeper.SetConnection(suite.chainB.GetContext(), proposedConnectionID, connectionEnd)
+				upgrade.Fields.ConnectionHops[0] = proposedConnectionID
+			},
+			connectiontypes.ErrInvalidConnectionState,
+		},
+		{
+			"proposed connection ends are not each other's counterparty",
+			func() {
+				// reuse existing connection to create a new connection in a non OPEN state
+				connectionEnd := path.EndpointB.GetConnection()
+				// ensure counterparty connectionID does not match connectionID set in counterparty proposed upgrade
+				connectionEnd.Counterparty.ConnectionId = "connection-50"
+
+				// set proposed connection in state
+				proposedConnectionID := "connection-100"
+				suite.chainB.GetSimApp().GetIBCKeeper().ConnectionKeeper.SetConnection(suite.chainB.GetContext(), proposedConnectionID, connectionEnd)
+				upgrade.Fields.ConnectionHops[0] = proposedConnectionID
+			},
+			types.ErrIncompatibleCounterpartyUpgrade,
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		suite.Run(tc.name, func() {
+			suite.SetupTest()
+
+			path = ibctesting.NewPath(suite.chainA, suite.chainB)
+			suite.coordinator.Setup(path)
+
+			upgradeVersion := fmt.Sprintf("%s-v2", mock.Version)
```

Comment:
could be worth adding another const to the mock pkg for this since its probably reused quite a bit. maybe `mock.UpgradeVersion` or something similar?

Reply:
opened #3734